### ISSUE TITLE
Refactor AlertService

### ISSUE THREAD

**Issue by baptiste-grob** [standardnotes/snjs#36]
I need to change the SNAlertService to support the addition of a blocking banner that shows a message, for when we upgrade the protocol version.
Because this will be a breaking change, I think it's a good opportunity to refactor the whole service.
Right now each platform (web/desktop & mobile) needs to re-implement the service methods so I think it would make more sense to only use an interface and require clients to give an implementation when constructing the SNApplication instance. Something like this:
The current file is here: https://github.com/standardnotes/snjs/blob/004/lib/services/alert_service.ts
And this is a basic version of where I'd like it to be:
```typescript
const enum ButtonType {
  Info = 0,
  Danger = 1,
}

type DismissBlockingDialog = () => void;

type AlertService = {
  confirm({
    text,
    title,
    confirmButtonText,
    confirmButtonType,
    cancelButtonText,
  }: {
    text: string,
    title?: string,
    confirmButtonText: string,
    confirmButtonType?: ButtonType,
    cancelButtonText?: string,
  }): Promise<boolean>;
  alert(text: string, title?: string, closeButtonText?: string): Promise<void>;
  blockingDialog(text: string): DismissBlockingDialog
}
```

I don't feel strongly about this proposed API, aside from two things:
- I need some way to instruct the client to show a blocking dialog and for SNJS to be able to dismiss it as well.
- I'd like for the `confirm` method to not `reject` when cancelling and instead resolve with a negative value. Otherwise it leads to awkward code like this:
```typescript
let didConfirm = true;
await this.alertService!.confirm(
  `Are you sure you want to delete ${itemsData.length} ${noun}?`
).catch(err => {
  didConfirm = false;
});
```

Which would look better like this:
```typescript
const didConfirm = await this.alertService!.confirm(
  `Are you sure you want to delete ${itemsData.length} ${noun}?`
);
```

What do you think? @mobitar @radko93

**Comment by mobitar** [comment]
`confirm` actually used to take an object as a param but I changed this to be a param list due to TS. But I'm fine with either approach. Refactoring it to remove objects was kind of labor intensive so to add it back again will probably take you 1-2 hours.

Regarding abstract implementation, we currently handle this via swapClasses inside of the application constructor. So seems like you would miss out on this functionality if you had to pass in an alertService separately.

You could have a separate `loadClasses` I suppose.

Just some thoughts, I don't have strong opinions on this either way.

**Comment by baptiste-grob** [comment]
If you think refactoring `confirm` would take a long time then I'm in favor of not addressing that right now and keep the signature the same, aside from not throwing on cancel.
Re swapClasses, I assume the idea is to let clients re-implement any class they want, however in the case of AlertService specifically I think it's misleading because it makes you think that you don't actually _need_ to re-implement it when really, you do. That's why I think it should be a hard requirement enforced by the type system.
I'll start working on this while heeding your warnings and being careful not to have it take too much time, IF @radko93 is okay with this API change 👍

**Comment by radko93** [comment]
I'm fine withe the API change 👍

**Comment by baptiste-grob** [comment]
Implemented in #41
